src/components/cl/hier/bcast/bcast_2step.c

```diff
@@ -0,0 +1,267 @@
+    }
+    return UCC_OK;
+}
+
+static inline ucc_rank_t
+find_root_net_rank(ucc_host_id_t root_host_id, ucc_cl_hier_team_t *cl_team)
+{
+    ucc_rank_t  net_rank  = UCC_RANK_INVALID;
+    ucc_sbgp_t *sbgp      = cl_team->sbgps[UCC_HIER_SBGP_NODE_LEADERS].sbgp;
+    ucc_team_t *core_team = cl_team->super.super.params.team;
+    ucc_rank_t  i, rank;
+
+    for (i = 0; i < sbgp->group_size; i++) {
+        rank = ucc_ep_map_eval(sbgp->map, i);
+        if (ucc_team_rank_host_id(rank, core_team) == root_host_id) {
+            net_rank = i;
+            break;
+        }
+    }
+    return net_rank;
+}
+
+static inline ucc_rank_t
+find_root_node_rank(ucc_rank_t root, ucc_cl_hier_team_t *cl_team)
+{
+    ucc_rank_t  node_rank  = UCC_RANK_INVALID;
+    ucc_sbgp_t *sbgp       = cl_team->sbgps[UCC_HIER_SBGP_NODE].sbgp;
+    ucc_rank_t  i;
+
+    for (i = 0; i < sbgp->group_size; i++) {
+        if (ucc_ep_map_eval(sbgp->map, i) == root) {
+            node_rank = i;
+            break;
+        }
+    }
+    return node_rank;
+}
+
+static ucc_status_t
+ucc_cl_hier_bcast_2step_init_schedule(ucc_base_coll_args_t *coll_args,
+                                        ucc_base_team_t *     team,
+                                        ucc_schedule_t **sched_p, int n_frags)
+{
+    ucc_cl_hier_team_t *cl_team = ucc_derived_of(team, ucc_cl_hier_team_t);
+    ucc_team_t         *core_team = team->params.team;
+    ucc_coll_task_t    *tasks[2] = {NULL, NULL};
+    int root                        = coll_args->args.root;
+    int rank                        = UCC_TL_TEAM_RANK(cl_team);
+    int root_on_local_node          = ucc_team_ranks_on_same_node(root, rank, core_team);
+
+    ucc_schedule_t      *schedule;
+    ucc_status_t         status;
+    ucc_base_coll_args_t args;
+    int                  n_tasks, i, first_task;
+
+    args           = *coll_args;
+    n_tasks        = 0;
+    first_task     = 0;
+
+    schedule = &ucc_cl_hier_get_schedule(cl_team)->super.super;
+    if (ucc_unlikely(!schedule)) {
+        return UCC_ERR_NO_MEMORY;
+    }
+    status         = ucc_schedule_init(schedule, &args, team);
+    if (ucc_unlikely(UCC_OK != status)) {
+        goto out;
+    }
+
+    if (n_frags > 1) {
+        args.max_frag_count =
+            ucc_buffer_block_count(args.args.src.info.count, n_frags, 0);
+        args.mask |= UCC_BASE_CARGS_MAX_FRAG_COUNT;
+    }
+
+    if (SBGP_ENABLED(cl_team, NODE_LEADERS)) {
+        args.args.root = find_root_net_rank(ucc_team_rank_host_id(root, core_team), cl_team);
+        status =
+            ucc_coll_init(SCORE_MAP(cl_team, NODE_LEADERS), &args, &tasks[n_tasks]);
+        if (ucc_unlikely(UCC_OK != status)) {
+            goto out;
+        }
+        n_tasks++;
+        if (root_on_local_node && (root != rank)) {
+            first_task = 1;
+        }
+    }
+
+    if (SBGP_ENABLED(cl_team, NODE)) {
+        args.args.root = root_on_local_node
+            ? find_root_node_rank(root, cl_team) : 0;
```

Comment:
is it guaranteed that node leader rank always has rank 0 in node group?

Reply:
good point. it is true now because parameter node_leader_rank_id is set to 0 during topo init. I changed the code to use that correct value.